src/sync/partial_sync.cpp

```diff
@@ -16,20 +16,109 @@
 //
 ////////////////////////////////////////////////////////////////////////////
 
-#include "sync/partial_sync.hpp"
-
-#include "impl/notification_wrapper.hpp"
-#include "impl/object_accessor_impl.hpp"
 #include "object_schema.hpp"
 #include "results.hpp"
 #include "shared_realm.hpp"
+#include "subscription_state.hpp"
+#include "impl/collection_change_builder.hpp"
+#include "impl/notification_wrapper.hpp"
+#include "impl/object_accessor_impl.hpp"
+#include "sync/partial_sync.hpp"
 #include "sync/sync_config.hpp"
+#include "sync/sync_session.hpp"
 
 #include <realm/util/scope_exit.hpp>
 
 namespace realm {
 namespace partial_sync {
 
+SubscriptionState create_or_update_subscription(Realm::Config config, SharedGroup &sg, realm::_impl::CollectionChangeBuilder &changes, Query &query, SubscriptionState previous_state) {
+
+    if (!config.sync_config || !config.sync_config->is_partial) {
+        return SubscriptionState::NOT_SUPPORTED;
+    }
+
+    // FIXME: Question: Is it problematic to do a write transaction here? Should we move it to a background thread`?
+
+    SubscriptionState old_partial_sync_state = previous_state;
+    SubscriptionState new_partial_sync_state = SubscriptionState::UNDEFINED;
+    std::string partial_sync_error_message = "";
+
+    // TODO: Determine how to create key. Right now we just used the serialized query (are there any
+    // disadvantages to that?). Would it make sense to hash it (to keep it short), but then we need
+    // to handle hash collisions as well.
+    std::string key = "SELECT * FROM XXX"; // Awaiting a release with query.get_description();
+
+    // TODO: It might change how the query is serialized. See https://realmio.slack.com/archives/C80PLGQ8Z/p1511797410000188
+    std::string serialized_query = "SELECT * FROM XXX"; // Awaiting a release with query.get_description();
+
+    LangBindHelper::promote_to_write(sg);
+    Group& group = _impl::SharedGroupFriend::get_group(sg);
+    // Check current state and create subscription if needed. Throw in an error is found:
+    TableRef table = ObjectStore::table_for_object_type(group, "__ResultSets");
+    size_t name_idx = table->get_descriptor()->get_column_index("name");
+    size_t query_idx = table->get_descriptor()->get_column_index("query");
+    size_t status_idx = table->get_descriptor()->get_column_index("status");
+    size_t error_idx = table->get_descriptor()->get_column_index("error_message");
+    size_t matches_property_idx = table->get_descriptor()->get_column_index("matches_property");
+    TableView results = (key == table->column<StringData>(name_idx)).find_all(0);
+    if (results.size() > 0) {
+        // Subscription with that ID already exist. Verify that we are not trying to reuse an
+        // existing name for a different query.
+        REALM_ASSERT(results.size() == 1);
+        auto obj = results.get(0);
+        if (obj.get_string(query_idx) != serialized_query) {
+            std::stringstream ss;
```

Comment:
Use util::format() rather than stringstream (it is both less verbose and generates significantly smaller compiled code).